records/2024-04-24.read-database-for-blazer.md

```diff
@@ -0,0 +1,49 @@
+# Read Only Database Replication for Blazer
+
+## Status
+
+**DRAFT**.
+
+## Context
+
+We had an incident where our production database nodes were being restarted unexpectedly due to a long running query that was executed manually in Blazer. This prompted an investigation into how to ensure blazer queries do not impact production. The conclusion was that the database would have to be replicated to a secondary read-only instance exclusive to blazer. 
+
+## Current State
+
+Currently blazer points to the read-only endpoint of the notify database. When running queries against this database, it is possible the query will cause one of the read instances to crash, prompting an automatic restart of the instance by AWS. There are still two other instances available when this happens, but it is possible that there will be a brief interruption in service since the notify application uses both the read and read/write proxy endpoints regularly. 
+
+Restrictions on who can use blazer in production have been created in attempt to prevent this from happening again.
+
+### Option 1 - Manual Logical Replication
+
+AWS Aurora RDS does not support automatic replication to a secondary database, but it can be configured manually. The steps would be as follows:
+
+1. Create a second database cluster
+2. Configure the primary database to do logical replication to the secondary
+3. Configure Blazer to only use the secondary database
+
+There are two incurred costs with this solution - first, there is an additional cost of ~$100-200 per month per environment to host the secondary database. Secondly, since the replication logic is manually administered, there is an additional cost of salary time in maintaining this. 
+
+### Option 2 - Redshift
+
+AWS Redshift is a managed data warehouse solution that could ingest our data for querying. It would likely replace Blazer entirely, and require learning a new tool to access/query the data. The steps to implement would be as follows:
+
+1. Subscribe to AWS Redshift
+2. Configure the primary database to stream to AWS Redshift
+3. Invest in training the team in Redshift querying
+
+Since Redshift is a fully managed service, there is likely a heavy cost associated with it. The absolute minimum cost is approximately $180 USD per month. There is also a substantial salary cost in training the team to use redshift over blazer.
+
+### Option 3 - Status Quo
+
+Since restrictions on who can use Blazer (restricted to Dev only), there have been no more incidents. Given that the original incident did not cause an outage, the case has been made that restricting usage is an acceptable alternative. The users that were previously using Blazer have been redirected to using AWS Quicksight, which is a better tool for BI anyway.
```

Comment:
```suggestion
Since restrictions on who can use Blazer (restricted to Dev only), there have been no more incidents. Given that the original incident did not cause an outage, the case has been made that restricting usage is an acceptable alternative. The users that were previously using Blazer have been redirected to using AWS QuickSight, which is a better tool for BI anyway.
```